syntax = "proto3";

import "google/protobuf/empty.proto";

package DataApi;

service FileBrowser {
    rpc GetFileList(FileListRequest) returns (FileList);
    rpc GetImageInfo(FileRequest) returns (ImageInfo);
    rpc OpenImage(FileRequest) returns (OpenImageResponse);
    rpc CloseImage(CloseFileRequest) returns (google.protobuf.Empty);
    rpc GetData(GetDataRequest) returns (DataResponse);
}

message FileListRequest {
    string directoryName = 1;
}

message FileList {
    string directoryName = 1;
    repeated DirectoryInfo subDirectories = 2;
    repeated FileInfo files = 3;
}

message DirectoryInfo {
    string name = 1;
    int64  numItems = 2;
    int64 date = 3;
}

message FileInfo {
    string name = 1;
    int64 size = 2;
    int64 date = 3;
}

message ImageInfo {
    string fileName = 1 ;
    FileType fileType = 2;
    repeated int32 dimensions = 3;

    message HeaderEntry {
        string key = 1;
        string value = 2;
        string comment = 3;
    }

    repeated HeaderEntry header = 4;
}

enum FileType {
    Unknown = 0;
    Fits = 1;
    Hdf5 = 2;
    Mhd = 3;
}

message FileRequest {
    string directoryName = 1;
    string fileName = 2;
    string hduName = 3;
    int32 hduNum = 4;
}

message OpenImageResponse {
    int32 fileId = 1;
}

message CloseFileRequest {
    int32 fileId = 1;
}

message GetDataRequest {
    int32 fileId = 1;
    // Region configuration for the image (Unused for now)
    int32 xMin = 2;
    int32 xMax = 3;
    int32 yMin = 4;
    int32 yMax = 5;
    int32 zMin = 6;
    int32 zMax = 7;
    // Down-sampling configuration (Unused for now)
    int32 xyMip = 8;
    int32 zMip = 9;

    enum DownSamplingMode {
        Mean = 0;
        Max = 1;
        Extrema = 2;
    }
    DownSamplingMode downSamplingMode = 10;
}

message DataResponse {
    bytes rawData = 1;
    float minValue = 2;
    float maxValue = 3;
    // TODO: Compression
}

